### PR TITLE
doc: suggestions for OSSL_PROVIDER_load_ex design document

### PR DESCRIPTION
**Pull request by mspncp**
**doc: suggestions for OSSL_PROVIDER_load_ex design document**
    
    Late review comments for pull request #21604, sort of.

**doc: change name of OSSL_PROVIDER_load_ex design document**
    
    Use dashes instead of underscores, to be more consistent with
    existing document names. And speaking of consistency, introduce
    a consistent name transformation, which will scale better when
    design documents start filling the folder ;-)
    
        OSSL_PROVIDER_load_ex -> ossl-provider-load-ex.md

### CODE REVIEW

**Comment by beldmit** [comment]
A doc-nit failure is relevant :(

File: doc/designs/ossl-provider-load-ex.md [hunk 1]
```diff
@@ -1,12 +1,12 @@
-Providers run-time configuration
-================================
+OSSL_PROVIDER_load_ex - activating providers with run-time configuration
+========================================================================
```
Comment (beldmit) [inline doc/designs/ossl-provider-load-ex.md:2]: Not sure. The document originally was intended to describe all the possible options, not only the new one.

File: doc/designs/ossl-provider-load-ex.md [hunk 2]
```diff
@@ -21,50 +21,49 @@ OSSL_PROVIDER *OSSL_PROVIDER_load_ex(OSSL_LIB_CTX *libctx, const char *name,
...
 
 Current design
 --------------
 
-When the provider is loaded in the current library context and activated, the
-currently loaded provider will be returned as the result of
-`OSSL_PROVIDER_load_ex` call.
+When the provider is already loaded an activated in the current library context, the
+`OSSL_PROVIDER_load_ex` call will simply return the active provider and the extra
+parameters will be ignored.
```
Comment (beldmit) [inline doc/designs/ossl-provider-load-ex.md:50]: "simply returns" instead of future tense?

File: doc/designs/ossl-provider-load-ex.md [hunk 3]
```diff
@@ -21,50 +21,49 @@ OSSL_PROVIDER *OSSL_PROVIDER_load_ex(OSSL_LIB_CTX *libctx, const char *name,
...
 
-When the provider is loaded in the current library context and NOT activated,
-the parameters provided int the `OSSL_PROVIDER_load_ex` call will have the
-preference.
+In all other cases, the extra parameters provided by the `OSSL_PROVIDER_load_ex` call
+will be applied and overwrite existing parameter values from the config file.
```
Comment (beldmit) [inline doc/designs/ossl-provider-load-ex.md:53]: Probably it's worth mentioning that the parameters from the configuration file will be completely ignored.

**Comment by beldmit** [comment]
@mspncp you provided much better wording than I did, and I mostly agree with your amendments

File: doc/designs/ossl-provider-load-ex.md (hunk 3 above)
Comment (mspncp) [reply doc/designs/ossl-provider-load-ex.md:53]: Oh, then I misunderstood. I thought the extra parameters would be merged with ones from the config file, overwriting existing value.

Comment (beldmit) [reply doc/designs/ossl-provider-load-ex.md:53]: No. All-or-nothing.

File: doc/designs/ossl-provider-load-ex.md (hunk 1 above)
Comment (mspncp) [reply doc/designs/ossl-provider-load-ex.md:2]: I don't quite follow. This here is just a design document, not the official documentation or a manual page. You proposed a new API, `OSSL_PROVIDER_load_ex` and this is the design document for it. Mentioning the alternatives serves as a rationale for your design decisions but it is not the goal of this document to document all existing options. I would expect those to be already documented in detail in our manpages ;-)

Comment (beldmit) [reply doc/designs/ossl-provider-load-ex.md:2]: Fair point, I agree, thanks!

**Comment by mspncp** [comment]
I did a rebase [without tree-changes](https://github.com/openssl/openssl/compare/c4f060afbbc784f03cb0d737a85ce5b4210dcfd3..40f4c757cb45393240102a3a5e30d77c6eaa8526) to autosquash my fixups.  This was my [last fixup](https://github.com/openssl/openssl/commit/c4f060afbbc784f03cb0d737a85ce5b4210dcfd3).

**Review by beldmit — APPROVED** [review]
LGTM if tests passed

**Comment by beldmit** [comment]
Ping @openssl/committers for the 2nd review

**Review by t8m — APPROVED** [review]

File: doc/designs/ossl-provider-load-ex.md [hunk 4]
```diff
@@ -21,50 +21,49 @@ OSSL_PROVIDER *OSSL_PROVIDER_load_ex(OSSL_LIB_CTX *libctx, const char *name,
...
-Many applications use PKCS#11 API with a specific drivers. OpenSSL PKCS#11
+Many applications use PKCS#11 API with specific drivers. OpenSSL PKCS#11
 provider <https://github.com/latchset/pkcs11-provider> also provides a set of
-tweaks usable in particular situations. So there are at least several scenarios
-I have in mind:
+tweaks usable in particular situations. So there are several scenarios for which
+the new API can be used:
 
 1. Configure a provider in the config file, activate on demand
 2. Load/activate a provider run-time with parameters
```
Comment (DDvO) [inline doc/designs/ossl-provider-load-ex.md:43]: A question: so far, the doc said that at least scenarios 1. and 2. are possible.
This PR suggests that these are the only ones. Is this correct?

Comment (beldmit) [reply doc/designs/ossl-provider-load-ex.md:43]: Yes. The combination of the config and app-defined parameters is not supported.

**Comment by beldmit** [comment]
Merged. Many thanks for the clarification of the document!

**Pull request closed without merging**